// @repo/platform/src/theming/index.ts
// Theming utilities for the platform
import { useState, useEffect, createContext, useContext } from 'react';
import { Theme, darkTheme, lightTheme, jazerTheme } from '@repo/design-system/themes';

// Theme context
interface ThemeContextType {
  theme: Theme;
  setTheme: (theme: Theme) => void;
  themeName: string;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

// Theme names
export type ThemeName = 'light' | 'dark' | 'jazer';

// Initial theme based on system preference
const getInitialTheme = (): Theme => {
  if (typeof window !== 'undefined') {
    const savedTheme = window.localStorage.getItem('theme') as ThemeName | null;
    if (savedTheme) {
      switch (savedTheme) {
        case 'dark': return darkTheme;
        case 'jazer': return jazerTheme;
        default: return lightTheme;
      }
    }
    
    // Check system preference
    if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
      return darkTheme;
    }
  }
  
  return lightTheme;
};

// Theme provider component
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setThemeState] = useState<Theme>(getInitialTheme());
  
  useEffect(() => {
    // Apply theme to document
    document.documentElement.setAttribute('data-theme', theme.name);
    
    // Store theme preference
    window.localStorage.setItem('theme', theme.name);
  }, [theme]);

  const setTheme = (newTheme: Theme) => {
    setThemeState(newTheme);
  };

  const toggleTheme = () => {
    setThemeState(prev => {
      if (prev.name === 'light') return darkTheme;
      if (prev.name === 'dark') return jazerTheme;
      return lightTheme;
    });
  };

  const themeName = theme.name as ThemeName;

  return (
    <ThemeContext.Provider value={{ theme, setTheme, themeName, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

// Hook to use theme
export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};

// Utility function to apply theme styles
export const applyTheme = (theme: Theme) => {
  return {
    '--color-primary-50': theme.colors.primary[50],
    '--color-primary-100': theme.colors.primary[100],
    '--color-primary-200': theme.colors.primary[200],
    '--color-primary-300': theme.colors.primary[300],
    '--color-primary-400': theme.colors.primary[400],
    '--color-primary-500': theme.colors.primary[500],
    '--color-primary-600': theme.colors.primary[600],
    '--color-primary-700': theme.colors.primary[700],
    '--color-primary-800': theme.colors.primary[800],
    '--color-primary-900': theme.colors.primary[900],
    // Add other themeable properties as needed
  } as React.CSSProperties;
};
