// @repo/design-system/src/motion/index.ts
// Motion tokens and utilities for the design system
export const motion = {
  duration: {
    shortest: '100ms',
    shorter: '150ms',
    short: '200ms',
    standard: '250ms',
    long: '300ms',
    longer: '350ms',
    longest: '400ms',
  },
  easing: {
    standard: 'cubic-bezier(0.4, 0.0, 0.2, 1)',
    decelerate: 'cubic-bezier(0.0, 0.0, 0.2, 1)',
    accelerate: 'cubic-bezier(0.4, 0.0, 1, 1)',
    sharp: 'cubic-bezier(0.4, 0.0, 0.6, 1)',
  },
  transforms: {
    scale: {
      enter: 'scale(0.95)',
      exit: 'scale(1.05)',
    },
    translate: {
      enter: 'translateY(20px)',
      exit: 'translateY(-20px)',
    },
  },
};

// Animation presets
export const animationPresets = {
  fadeIn: {
    animation: 'fadeIn 150ms ease-out',
  },
  fadeOut: {
    animation: 'fadeOut 150ms ease-in',
  },
  slideIn: {
    animation: 'slideIn 200ms cubic-bezier(0.4, 0.0, 0.2, 1)',
  },
  slideOut: {
    animation: 'slideOut 200ms cubic-bezier(0.4, 0.0, 0.2, 1)',
  },
  scaleIn: {
    animation: 'scaleIn 200ms cubic-bezier(0.4, 0.0, 0.2, 1)',
  },
  scaleOut: {
    animation: 'scaleOut 200ms cubic-bezier(0.4, 0.0, 0.2, 1)',
  },
};

// CSS keyframes for animation presets
export const keyframes = `
  @keyframes fadeIn {
    from { opacity: 0; }
    to { opacity: 1; }
  }
  
  @keyframes fadeOut {
    from { opacity: 1; }
    to { opacity: 0; }
  }
  
  @keyframes slideIn {
    from { transform: translateY(20px); opacity: 0; }
    to { transform: translateY(0); opacity: 1; }
  }
  
  @keyframes slideOut {
    from { transform: translateY(0); opacity: 1; }
    to { transform: translateY(-20px); opacity: 0; }
  }
  
  @keyframes scaleIn {
    from { transform: scale(0.95); opacity: 0; }
    to { transform: scale(1); opacity: 1; }
  }
  
  @keyframes scaleOut {
    from { transform: scale(1); opacity: 1; }
    to { transform: scale(0.95); opacity: 0; }
  }
`;
