// @repo/platform/src/navigation/index.ts
// Navigation utilities for the platform
import { useState, useEffect } from 'react';
import { navigateTo } from '../routing';

// Navigation item interface
export interface NavItem {
  id: string;
  title: string;
  path: string;
  icon?: React.ReactNode;
  children?: NavItem[];
  requiresAuth?: boolean;
}

// Navigation state
interface NavigationState {
  currentPath: string;
  breadcrumbs: NavItem[];
  sidebarOpen: boolean;
}

// Hook for navigation state management
export const useNavigation = (navItems: NavItem[]) => {
  const [state, setState] = useState<NavigationState>({
    currentPath: typeof window !== 'undefined' ? window.location.pathname : '/',
    breadcrumbs: [],
    sidebarOpen: false,
  });

  // Update breadcrumbs based on current path
  useEffect(() => {
    if (typeof window !== 'undefined') {
      const path = window.location.pathname;
      setState(prev => ({
        ...prev,
        currentPath: path,
        breadcrumbs: generateBreadcrumbs(path, navItems),
      }));
    }
  }, [navItems]);

  const generateBreadcrumbs = (path: string, items: NavItem[]): NavItem[] => {
    // Find the path in the navigation tree
    const findPath = (items: NavItem[], targetPath: string): NavItem[] => {
      for (const item of items) {
        if (item.path === targetPath) {
          return [item];
        }
        
        if (item.children) {
          const childPath = findPath(item.children, targetPath);
          if (childPath.length) {
            return [item, ...childPath];
          }
        }
      }
      return [];
    };

    return findPath(items, path);
  };

  const toggleSidebar = () => {
    setState(prev => ({
      ...prev,
      sidebarOpen: !prev.sidebarOpen,
    }));
  };

  const closeSidebar = () => {
    setState(prev => ({
      ...prev,
      sidebarOpen: false,
    }));
  };

  return {
    ...state,
    toggleSidebar,
    closeSidebar,
    navigateTo,
  };
};

// Navigation helpers
export const isActivePath = (currentPath: string, targetPath: string) => {
  return currentPath === targetPath || currentPath.startsWith(`${targetPath}/`);
};
