// @repo/platform/src/storage/index.ts
// Storage utilities for the platform
import { useState, useEffect } from 'react';

// Storage type
export type StorageType = 'localStorage' | 'sessionStorage';

// Storage utility class
class StorageUtil {
  private storage: Storage;

  constructor(type: StorageType = 'localStorage') {
    this.storage = type === 'localStorage' 
      ? typeof window !== 'undefined' ? window.localStorage : null 
      : typeof window !== 'undefined' ? window.sessionStorage : null;
  }

  // Get item from storage
  getItem<T>(key: string, defaultValue?: T): T | null {
    if (!this.storage) return defaultValue || null;
    
    try {
      const item = this.storage.getItem(key);
      return item ? JSON.parse(item) : defaultValue || null;
    } catch (error) {
      console.warn(`Error reading from ${this.storage}:`, error);
      return defaultValue || null;
    }
  }

  // Set item in storage
  setItem<T>(key: string, value: T): void {
    if (!this.storage) return;
    
    try {
      this.storage.setItem(key, JSON.stringify(value));
    } catch (error) {
      console.warn(`Error writing to ${this.storage}:`, error);
    }
  }

  // Remove item from storage
  removeItem(key: string): void {
    if (!this.storage) return;
    
    try {
      this.storage.removeItem(key);
    } catch (error) {
      console.warn(`Error removing from ${this.storage}:`, error);
    }
  }

  // Clear all items from storage
  clear(): void {
    if (!this.storage) return;
    
    try {
      this.storage.clear();
    } catch (error) {
      console.warn(`Error clearing ${this.storage}:`, error);
    }
  }

  // Check if storage is available
  isAvailable(): boolean {
    return this.storage !== null;
  }
}

// Global storage instances
export const localStorageUtil = new StorageUtil('localStorage');
export const sessionStorageUtil = new StorageUtil('sessionStorage');

// Hook for local storage
export const useLocalStorage = <T>(key: string, initialValue: T) => {
  const [storedValue, setStoredValue] = useState<T>(() => {
    try {
      const item = localStorageUtil.getItem<T>(key);
      return item !== null ? item : initialValue;
    } catch (error) {
      console.warn(`Error reading from localStorage:`, error);
      return initialValue;
    }
  });

  const setValue = (value: T | ((val: T) => T)) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);
      localStorageUtil.setItem(key, valueToStore);
    } catch (error) {
      console.warn(`Error setting to localStorage:`, error);
    }
  };

  return [storedValue, setValue] as const;
};

// Hook for session storage
export const useSessionStorage = <T>(key: string, initialValue: T) => {
  const [storedValue, setStoredValue] = useState<T>(() => {
    try {
      const item = sessionStorageUtil.getItem<T>(key);
      return item !== null ? item : initialValue;
    } catch (error) {
      console.warn(`Error reading from sessionStorage:`, error);
      return initialValue;
    }
  });

  const setValue = (value: T | ((val: T) => T)) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);
      sessionStorageUtil.setItem(key, valueToStore);
    } catch (error) {
      console.warn(`Error setting to sessionStorage:`, error);
    }
  };

  return [storedValue, setValue] as const;
};

// Hook for storage event
export const useStorageListener = (callback: (e: StorageEvent) => void) => {
  useEffect(() => {
    const handleStorageChange = (e: StorageEvent) => {
      callback(e);
    };

    window.addEventListener('storage', handleStorageChange);

    return () => {
      window.removeEventListener('storage', handleStorageChange);
    };
  }, [callback]);
};
