// @repo/design-system/src/themes/index.ts
// Theme definitions for the design system
import { colors, spacing, typography } from '../tokens';

export interface Theme {
  colors: typeof colors;
  spacing: typeof spacing;
  typography: typeof typography;
  name: string;
}

// Base theme with standard values
export const baseTheme: Theme = {
  name: 'base',
  colors,
  spacing,
  typography,
};

// Dark theme extending base theme
export const darkTheme: Theme = {
  ...baseTheme,
  name: 'dark',
  colors: {
    ...colors,
    primary: {
      50: '#1e293b',
      100: '#334155',
      200: '#475569',
      300: '#64748b',
      400: '#94a3b8',
      500: '#cbd5e1',
      600: '#e2e8f0',
      700: '#f1f5f9',
      800: '#f8fafc',
      900: '#ffffff',
    },
  },
};

// Light theme with standard values
export const lightTheme: Theme = {
  ...baseTheme,
  name: 'light',
};

// JaZeR theme based on the original design
export const jazerTheme: Theme = {
  ...baseTheme,
  name: 'jazer',
  colors: {
    ...colors,
    primary: {
      50: '#050505',  // Deepest void
      100: '#0A0A0A', // Primary card background
      200: '#121212', // Secondary/Hover states
      300: '#1e293b',
      400: '#334155',
      500: '#475569',
      600: '#64748b',
      700: '#94a3b8',
      800: '#cbd5e1',
      900: '#e2e8f0',
    },
    secondary: {
      ...colors.secondary,
      500: '#00F0FF', // JaZeR accent color
    },
  },
};
