// @repo/design-system/src/elevation/index.ts
// Elevation tokens for the design system (shadows and depth)
export const elevation = {
  none: 'none',
  xs: '0 1px 2px 0 rgba(0, 0, 0, 0.05)',
  sm: '0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px -1px rgba(0, 0, 0, 0.1)',
  md: '0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -2px rgba(0, 0, 0, 0.1)',
  lg: '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -4px rgba(0, 0, 0, 0.1)',
  xl: '0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 8px 10px -6px rgba(0, 0, 0, 0.1)',
  '2xl': '0 25px 50px -12px rgba(0, 0, 0, 0.25)',
};

// Glassmorphism effect (for Apple Liquid Glass design)
export const glass = {
  standard: `
    background: rgba(255, 255, 255, 0.1);
    backdrop-filter: blur(10px);
    -webkit-backdrop-filter: blur(10px);
    border: 1px solid rgba(255, 255, 255, 0.18);
    box-shadow: 0 8px 32px 0 rgba(31, 38, 135, 0.37);
  `,
  dark: `
    background: rgba(0, 0, 0, 0.2);
    backdrop-filter: blur(10px);
    -webkit-backdrop-filter: blur(10px);
    border: 1px solid rgba(255, 255, 255, 0.05);
    box-shadow: 0 8px 32px 0 rgba(31, 38, 135, 0.15);
  `,
  jazer: `
    background: rgba(10, 10, 10, 0.7);  /* Frosted inputs/panels */
    backdrop-filter: blur(12px);
    -webkit-backdrop-filter: blur(12px);
    border: 1px solid rgba(255, 255, 255, 0.1);
    box-shadow: 0 4px 20px rgba(0, 0, 0, 0.25);
  `,
};

// Elevation presets based on common use cases
export const elevationPresets = {
  card: elevation.md,
  modal: elevation.xl,
  dropdown: elevation.lg,
  tooltip: elevation.sm,
  navigation: elevation.md,
  floatingButton: elevation.lg,
};

// Function to apply elevation preset
export const applyElevation = (presetName: keyof typeof elevationPresets) => {
  return { boxShadow: elevationPresets[presetName] };
};
