// @repo/platform/src/telemetry/index.ts
// Telemetry utilities for the platform (stubs only - no network calls)
export interface TelemetryEvent {
  name: string;
  timestamp: number;
  properties?: Record<string, any>;
  userId?: string;
  sessionId?: string;
}

// Telemetry API interface
export interface TelemetryAPI {
  log(event: TelemetryEvent): void;
  logError(error: Error, context?: Record<string, any>): void;
  logPerformance(metric: string, value: number, context?: Record<string, any>): void;
  flush(): void;
}

// Stub implementation of telemetry
class StubTelemetry implements TelemetryAPI {
  private eventQueue: TelemetryEvent[] = [];

  log(event: TelemetryEvent): void {
    // In a real implementation, this would send data to a telemetry service
    // For now, we're just queueing events and logging in development mode
    this.eventQueue.push(event);
    
    if (process.env.NODE_ENV === 'development') {
      console.log('[Telemetry Log]', event);
    }
  }

  logError(error: Error, context?: Record<string, any>): void {
    const event: TelemetryEvent = {
      name: 'error',
      timestamp: Date.now(),
      properties: {
        message: error.message,
        stack: error.stack,
        ...context,
      },
    };
    
    this.log(event);
  }

  logPerformance(metric: string, value: number, context?: Record<string, any>): void {
    const event: TelemetryEvent = {
      name: 'performance',
      timestamp: Date.now(),
      properties: {
        metric,
        value,
        ...context,
      },
    };
    
    this.log(event);
  }

  flush(): void {
    // In a real implementation, this would send queued events to a service
    if (process.env.NODE_ENV === 'development' && this.eventQueue.length > 0) {
      console.log('[Telemetry Flush]', this.eventQueue);
    }
    this.eventQueue = [];
  }
}

// Global telemetry instance
const telemetry: TelemetryAPI = new StubTelemetry();

// Export the telemetry instance
export { telemetry };

// Convenience functions
export const logEvent = (name: string, properties?: Record<string, any>) => {
  telemetry.log({ 
    name, 
    timestamp: Date.now(), 
    properties 
  });
};

export const logError = (error: Error, context?: Record<string, any>) => {
  telemetry.logError(error, context);
};

export const logPerformance = (metric: string, value: number, context?: Record<string, any>) => {
  telemetry.logPerformance(metric, value, context);
};
