// @repo/platform/src/routing/index.ts
// Routing utilities for the platform
import { useRouter } from 'next/router';

// Type for route parameters
export interface RouteParams {
  [key: string]: string | string[];
}

// Route configuration type
export interface RouteConfig {
  path: string;
  name: string;
  component?: React.ComponentType;
  exact?: boolean;
  requiresAuth?: boolean;
}

// Navigation functions
export const navigateTo = (path: string, params?: RouteParams) => {
  const router = useRouter();
  if (params) {
    const queryString = new URLSearchParams(params as Record<string, string>).toString();
    return router.push(`${path}${queryString ? `?${queryString}` : ''}`);
  }
  return router.push(path);
};

// Get current route parameters
export const useRouteParams = () => {
  const router = useRouter();
  return router.query as RouteParams;
};

// Get current route path
export const useCurrentPath = () => {
  const router = useRouter();
  return router.pathname;
};

// Check if current route matches a pattern
export const useIsRoute = (pattern: string) => {
  const currentPath = useCurrentPath();
  return currentPath === pattern || currentPath.startsWith(pattern);
};
