// @repo/platform/src/responsive/index.ts
// Responsive utilities for the platform
import { useState, useEffect } from 'react';

// Breakpoints
export const BREAKPOINTS = {
  sm: 640,   // 640px
  md: 768,   // 768px
  lg: 1024,  // 1024px
  xl: 1280,  // 1280px
  '2xl': 1536, // 1536px
};

// Breakpoint names
export type Breakpoint = keyof typeof BREAKPOINTS;

// Responsive hook
export const useResponsive = () => {
  const [breakpoint, setBreakpoint] = useState<Breakpoint>('md');
  const [windowSize, setWindowSize] = useState({
    width: typeof window !== 'undefined' ? window.innerWidth : 0,
    height: typeof window !== 'undefined' ? window.innerHeight : 0,
  });

  useEffect(() => {
    if (typeof window === 'undefined') return;

    const handleResize = () => {
      const width = window.innerWidth;
      setWindowSize({ width, height: window.innerHeight });

      // Determine current breakpoint
      if (width < BREAKPOINTS.sm) {
        setBreakpoint('sm');
      } else if (width < BREAKPOINTS.md) {
        setBreakpoint('md');
      } else if (width < BREAKPOINTS.lg) {
        setBreakpoint('lg');
      } else if (width < BREAKPOINTS.xl) {
        setBreakpoint('xl');
      } else {
        setBreakpoint('2xl');
      }
    };

    // Initial call
    handleResize();

    // Add event listener
    window.addEventListener('resize', handleResize);

    // Cleanup
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  // Helper functions
  const isMobile = breakpoint === 'sm';
  const isTablet = breakpoint === 'md';
  const isDesktop = breakpoint === 'lg' || breakpoint === 'xl' || breakpoint === '2xl';
  const isLargeScreen = breakpoint === 'xl' || breakpoint === '2xl';

  // Check if current breakpoint matches or is larger than given breakpoint
  const isGreaterOrEqual = (bp: Breakpoint) => {
    const currentIdx = Object.keys(BREAKPOINTS).indexOf(breakpoint);
    const targetIdx = Object.keys(BREAKPOINTS).indexOf(bp);
    return currentIdx >= targetIdx;
  };

  // Check if current breakpoint matches or is smaller than given breakpoint
  const isLessOrEqual = (bp: Breakpoint) => {
    const currentIdx = Object.keys(BREAKPOINTS).indexOf(breakpoint);
    const targetIdx = Object.keys(BREAKPOINTS).indexOf(bp);
    return currentIdx <= targetIdx;
  };

  return {
    breakpoint,
    windowSize,
    isMobile,
    isTablet,
    isDesktop,
    isLargeScreen,
    isGreaterOrEqual,
    isLessOrEqual,
  };
};

// Hook to conditionally render based on breakpoint
export const useBreakpoint = (bp: Breakpoint) => {
  const { breakpoint: currentBp } = useResponsive();
  return currentBp === bp;
};

// Hook for mobile detection
export const useIsMobile = () => {
  const { isMobile } = useResponsive();
  return isMobile;
};

// Hook for desktop detection
export const useIsDesktop = () => {
  const { isDesktop } = useResponsive();
  return isDesktop;
};
