// @repo/content/src/manifest/index.ts
// Content manifest utilities for the platform
export interface ContentManifest {
  id: string;
  name: string;
  version: string;
  description: string;
  sourcePath: string;
  files: ContentFile[];
  dependencies?: string[];
  createdAt: Date;
  updatedAt: Date;
}

export interface ContentFile {
  id: string;
  name: string;
  path: string;
  type: string;
  size: number;
  hash: string;
  metadata?: Record<string, any>;
}

// Create a content manifest from source files
export const createContentManifest = async (
  id: string,
  name: string,
  sourcePath: string,
  files: { path: string; name: string; type: string; metadata?: Record<string, any> }[]
): Promise<ContentManifest> => {
  // In a real implementation, this would read file sizes and calculate hashes
  // For now, we'll simulate with placeholder values
  const manifestFiles: ContentFile[] = files.map((file, index) => ({
    id: `file-${index}`,
    name: file.name,
    path: file.path,
    type: file.type,
    size: Math.floor(Math.random() * 10000) + 1000, // Random size between 1KB-10KB
    hash: `hash-${index}`, // Placeholder hash
    metadata: file.metadata,
  }));

  return {
    id,
    name,
    version: '1.0.0',
    description: `Manifest for ${name}`,
    sourcePath,
    files: manifestFiles,
    createdAt: new Date(),
    updatedAt: new Date(),
  };
};

// Load a content manifest from a JSON file
export const loadContentManifest = async (manifestPath: string): Promise<ContentManifest> => {
  // In a real implementation, this would read the manifest file from disk
  // For now, we'll return a placeholder manifest
  console.warn(`Loading content manifest from ${manifestPath} - placeholder implementation`);
  
  return {
    id: 'placeholder-manifest',
    name: 'Placeholder Manifest',
    version: '1.0.0',
    description: 'Placeholder manifest for demonstration',
    sourcePath: manifestPath,
    files: [],
    createdAt: new Date(),
    updatedAt: new Date(),
  };
};

// Validate a content manifest
export const validateContentManifest = (manifest: ContentManifest): boolean => {
  const requiredFields: (keyof ContentManifest)[] = [
    'id', 'name', 'version', 'description', 'sourcePath', 'files'
  ];

  for (const field of requiredFields) {
    if (!manifest[field]) {
      console.error(`Missing required field in manifest: ${field}`);
      return false;
    }
  }

  // Validate files
  if (!Array.isArray(manifest.files) || manifest.files.length === 0) {
    console.error('Manifest must contain at least one file');
    return false;
  }

  for (const file of manifest.files) {
    if (!file.id || !file.name || !file.path || !file.type) {
      console.error('Each file in manifest must have id, name, path, and type');
      return false;
    }
  }

  return true;
};

// Update a content manifest
export const updateContentManifest = (
  manifest: ContentManifest,
  updates: Partial<ContentManifest>
): ContentManifest => {
  return {
    ...manifest,
    ...updates,
    updatedAt: new Date(),
  };
};
