// @repo/platform/src/analytics/index.ts
// Analytics utilities for the platform (stubs only - no tracking by default)
export interface AnalyticsEvent {
  name: string;
  properties?: Record<string, any>;
}

export interface AnalyticsUser {
  id?: string;
  traits?: Record<string, any>;
}

// Analytics API interface
export interface AnalyticsAPI {
  track(event: AnalyticsEvent): void;
  identify(user: AnalyticsUser): void;
  page(name: string, properties?: Record<string, any>): void;
  flush(): void;
  reset(): void;
}

// Stub implementation of analytics
class StubAnalytics implements AnalyticsAPI {
  track(event: AnalyticsEvent): void {
    // In a real implementation, this would send data to an analytics service
    // For now, we're just logging to console in development mode
    if (process.env.NODE_ENV === 'development') {
      console.log('[Analytics Track]', event);
    }
  }

  identify(user: AnalyticsUser): void {
    // In a real implementation, this would identify the user to an analytics service
    if (process.env.NODE_ENV === 'development') {
      console.log('[Analytics Identify]', user);
    }
  }

  page(name: string, properties?: Record<string, any>): void {
    // In a real implementation, this would track a page view
    if (process.env.NODE_ENV === 'development') {
      console.log('[Analytics Page]', name, properties);
    }
  }

  flush(): void {
    // In a real implementation, this would flush any pending events
  }

  reset(): void {
    // In a real implementation, this would reset the analytics state
  }
}

// Global analytics instance
const analytics: AnalyticsAPI = new StubAnalytics();

// Export the analytics instance
export { analytics };

// Convenience functions
export const trackEvent = (name: string, properties?: Record<string, any>) => {
  analytics.track({ name, properties });
};

export const identifyUser = (user: AnalyticsUser) => {
  analytics.identify(user);
};

export const trackPage = (name: string, properties?: Record<string, any> = {}) => {
  analytics.page(name, properties);
};
