// @repo/ui/src/components/Card.tsx
import React from 'react';
import { colors, spacing, typography } from '@repo/design-system';
import { motion } from '@repo/design-system/motion';
import { elevation, glass } from '@repo/design-system/elevation';

interface CardProps {
  children: React.ReactNode;
  variant?: 'standard' | 'glass' | 'elevated';
  className?: string;
  style?: React.CSSProperties;
}

export const Card: React.FC<CardProps> = ({
  children,
  variant = 'standard',
  className = '',
  style = {},
}) => {
  const baseClasses = 'rounded-xl border';

  const variantStyles = {
    standard: {
      className: 'bg-white border-gray-200',
      style: { boxShadow: elevation.md },
    },
    glass: {
      className: 'border-white/10',
      style: {
        ...glass.standard,
        ...style,
      },
    },
    elevated: {
      className: 'bg-white border-gray-200',
      style: { boxShadow: elevation.lg },
    },
  };

  const { className: variantClass, style: variantStyle } = variantStyles[variant];

  return (
    <div
      className={`${baseClasses} ${variantClass} ${className}`}
      style={{
        ...variantStyle,
        transition: `all ${motion.duration.standard} ${motion.easing.standard}`,
        ...style,
      }}
    >
      {children}
    </div>
  );
};
