// @repo/ui/src/components/Navigation.tsx
import React from 'react';
import { colors, spacing, typography } from '@repo/design-system';
import { motion } from '@repo/design-system/motion';
import { elevation } from '@repo/design-system/elevation';

interface NavItem {
  id: string;
  title: string;
  href?: string;
  onClick?: () => void;
  icon?: React.ReactNode;
  badge?: string | number;
}

interface NavigationProps {
  items: NavItem[];
  currentPath?: string;
  variant?: 'horizontal' | 'vertical';
  className?: string;
}

export const Navigation: React.FC<NavigationProps> = ({
  items,
  currentPath = '',
  variant = 'horizontal',
  className = '',
}) => {
  const orientation = variant === 'horizontal' 
    ? 'flex-row items-center space-x-1' 
    : 'flex-col items-stretch space-y-1 p-2';

  return (
    <nav 
      className={`
        ${variant === 'horizontal' 
          ? 'bg-white border-b border-gray-200' 
          : 'bg-white border-r border-gray-200'}
        ${orientation}
        rounded-lg
        ${className}
      `}
      style={{
        boxShadow: variant === 'horizontal' ? elevation.sm : 'none',
        transition: `all ${motion.duration.standard} ${motion.easing.standard}`,
      }}
    >
      {items.map((item) => {
        const isActive = currentPath === item.href;
        const baseLinkClasses = `
          flex items-center gap-2 px-4 py-2.5 rounded-md text-sm font-medium
          transition-colors ${motion.duration.standard} ${motion.easing.standard}
          ${isActive 
            ? 'bg-blue-100 text-blue-700' 
            : 'text-gray-600 hover:bg-gray-100 hover:text-gray-900'}
        `;
        
        return (
          <a
            key={item.id}
            href={item.href}
            onClick={(e) => {
              if (item.onClick) {
                e.preventDefault();
                item.onClick();
              }
            }}
            className={baseLinkClasses}
          >
            {item.icon && <span>{item.icon}</span>}
            <span>{item.title}</span>
            {item.badge && (
              <span className="ml-auto inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-blue-100 text-blue-800">
                {item.badge}
              </span>
            )}
          </a>
        );
      })}
    </nav>
  );
};
