// @repo/ui/src/components/Button.tsx
import React from 'react';
import { colors, spacing, typography } from '@repo/design-system';
import { motion } from '@repo/design-system/motion';
import { elevation } from '@repo/design-system/elevation';

interface ButtonProps {
  children: React.ReactNode;
  variant?: 'primary' | 'secondary' | 'ghost' | 'outline';
  size?: 'sm' | 'md' | 'lg';
  disabled?: boolean;
  onClick?: () => void;
  className?: string;
}

export const Button: React.FC<ButtonProps> = ({
  children,
  variant = 'primary',
  size = 'md',
  disabled = false,
  onClick,
  className = '',
}) => {
  // Define CSS classes based on props
  const baseClasses = `inline-flex items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 ${
    disabled ? 'opacity-50 cursor-not-allowed' : ''
  }`;

  const variantClasses = {
    primary: `bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500`,
    secondary: `bg-gray-200 text-gray-900 hover:bg-gray-300 focus:ring-gray-500`,
    ghost: `bg-transparent text-gray-700 hover:bg-gray-100 focus:ring-gray-500`,
    outline: `border border-gray-300 bg-transparent text-gray-700 hover:bg-gray-50 focus:ring-gray-500`,
  };

  const sizeClasses = {
    sm: `text-sm px-3 py-1.5`,
    md: `text-base px-4 py-2`,
    lg: `text-lg px-6 py-3`,
  };

  const classes = `${baseClasses} ${variantClasses[variant]} ${sizeClasses[size]} ${className}`;

  return (
    <button
      className={classes}
      onClick={onClick}
      disabled={disabled}
      style={{
        transition: `all ${motion.duration.standard} ${motion.easing.standard}`,
      }}
    >
      {children}
    </button>
  );
};
