// @repo/ui/src/components/Table.tsx
import React from 'react';
import { colors, spacing, typography } from '@repo/design-system';
import { motion } from '@repo/design-system/motion';
import { elevation } from '@repo/design-system/elevation';

interface TableColumn<T> {
  key: string;
  title: string;
  render?: (item: T) => React.ReactNode;
  width?: string;
  align?: 'left' | 'center' | 'right';
}

interface TableProps<T> {
  data: T[];
  columns: TableColumn<T>[];
  rowKey: (item: T) => string | number;
  onRowClick?: (item: T) => void;
  className?: string;
  size?: 'sm' | 'md' | 'lg';
}

export const Table = <T extends Record<string, any>>({
  data,
  columns,
  rowKey,
  onRowClick,
  className = '',
  size = 'md',
}: TableProps<T>) => {
  const sizeClasses = {
    sm: 'text-sm',
    md: 'text-base',
    lg: 'text-lg',
  };

  return (
    <div className={`overflow-hidden rounded-lg border border-gray-200 ${className}`}>
      <table className="min-w-full divide-y divide-gray-200">
        <thead className="bg-gray-50">
          <tr>
            {columns.map((column) => (
              <th
                key={column.key}
                scope="col"
                className={`
                  px-6 py-3 text-left font-medium text-gray-500 uppercase tracking-wider
                  ${sizeClasses[size]}
                `}
                style={{ width: column.width }}
              >
                {column.title}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="bg-white divide-y divide-gray-200">
          {data.map((item) => (
            <tr
              key={rowKey(item)}
              onClick={() => onRowClick && onRowClick(item)}
              className={`
                ${onRowClick ? 'cursor-pointer hover:bg-gray-50' : ''}
                transition-colors ${motion.duration.shorter} ${motion.easing.standard}
              `}
            >
              {columns.map((column) => (
                <td
                  key={`${rowKey(item)}-${column.key}`}
                  className={`
                    px-6 py-4 whitespace-nowrap
                    ${sizeClasses[size]} text-gray-900
                    ${column.align ? `text-${column.align}` : ''}
                  `}
                >
                  {column.render ? column.render(item) : item[column.key]}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
