// @repo/design-system/src/typography/index.ts
// Typography utilities for the design system
import { typography as tokenTypography } from '../tokens';

export const font = tokenTypography.fontFamily;

export const fontSize = tokenTypography.fontSize;

export const fontWeight = tokenTypography.fontWeight;

// Typography scale presets
export const typographyPresets = {
  heading1: {
    fontSize: fontSize['5xl'],
    fontWeight: fontWeight.bold,
    lineHeight: '1.2',
  },
  heading2: {
    fontSize: fontSize['4xl'],
    fontWeight: fontWeight.bold,
    lineHeight: '1.25',
  },
  heading3: {
    fontSize: fontSize['3xl'],
    fontWeight: fontWeight.semibold,
    lineHeight: '1.3',
  },
  body: {
    fontSize: fontSize.base,
    fontWeight: fontWeight.normal,
    lineHeight: '1.5',
  },
  caption: {
    fontSize: fontSize.sm,
    fontWeight: fontWeight.medium,
    lineHeight: '1.4',
  },
};

// Utility function to apply typography preset
export const applyPreset = (presetName: keyof typeof typographyPresets) => {
  return typographyPresets[presetName];
};
